File: selected_coding_interview/codes/java/lc_232_implement_queue_using_stacks/lc_232_implement_queue_using_stacks.java
/*
* File: lc_232_implement_queue_using_stacks.java
* Created Time: 2025-12-30
* Author: krahets
*/

package lc_232_implement_queue_using_stacks;

import include.*;
import java.util.*;

// ===== Solution Code =====
    class MyQueue {
        private Stack<Integer> A;
        private Stack<Integer> B;
        public MyQueue() {
            A = new Stack<>();
            B = new Stack<>();
        }
        public void push(int x) {
            A.push(x);
        }
        public int pop() {
            int peek = peek();
            B.pop();
            return peek;
        }
        public int peek() {
            if (!B.isEmpty()) return B.peek();
            if (A.isEmpty()) return -1;
            while (!A.isEmpty()){
                B.push(A.pop());
            }
            return B.peek();
        }
        public boolean empty() {
            return A.isEmpty() && B.isEmpty();
        }
    }

public class lc_232_implement_queue_using_stacks {
    public static void main(String[] args) {
        // ======= Test Case =======
        // Test case 1
        //operations = ["MyQueue", "push", "push", "peek", "pop", "empty"]
        //values = [[], [1], [2], [], [], []]

        // ====== Driver Code ======
        MyQueue obj = new MyQueue();
        obj.push(1);
        obj.push(2);
        System.out.println("Peek:" + " " + obj.peek());
        System.out.println("Pop:" + " " + obj.pop());
        System.out.println("Empty:" + " " + obj.empty());

    }
}

File: selected_coding_interview/codes/python/lc_232_implement_queue_using_stacks.py
"""
File: lc_232_implement_queue_using_stacks.py
Created Time: 2025-12-30
Author: krahets
"""

from include import *


# ===== Solution Code =====
class MyQueue:
    def __init__(self):
        self.A, self.B = [], []
    def push(self, x: int) -> None:
        self.A.append(x)
    def pop(self) -> int:
        peek = self.peek()
        self.B.pop()
        return peek
    def peek(self) -> int:
        if self.B: return self.B[-1]
        if not self.A: return -1
        # 将栈 A 的元素依次移动至栈 B
        while self.A:
            self.B.append(self.A.pop())
        return self.B[-1]
    def empty(self) -> bool:
        return not self.A and not self.B


# ======= Test Case =======
# Test case 1
operations = ["MyQueue", "push", "push", "peek", "pop", "empty"]
values = [[], [1], [2], [], [], []]

# ====== Driver Code ======
obj = MyQueue()
obj.push(1)
obj.push(2)
print("Peek:", obj.peek())
print("Pop:", obj.pop())
print("Empty:", obj.empty())
